op-program/client/l2/engine.go

```diff
@@ -70,6 +76,10 @@ func (o *OracleEngine) GetPayload(ctx context.Context, payloadInfo eth.PayloadIn
 }
 
 func (o *OracleEngine) ForkchoiceUpdate(ctx context.Context, state *eth.ForkchoiceState, attr *eth.PayloadAttributes) (*eth.ForkchoiceUpdatedResult, error) {
+	header := o.backend.GetHeaderByHash(state.SafeBlockHash)
+
+	o.hintWriter.Hint(ExecutionWitnessHint(header.Number.Uint64() + 1))
```

Comment:
Good question - InsertUnsafePayload and NewPayload aren't called by op-program before block processing begins.

Reply:
We probably should be sending the hint in both places since we might start processing a block as part of converting a batch to a block in `forkChoiceUpdated` but may also start processing a block in a `newPayload` call.  It happens that currently the build always happens via `forkChoiceUpdated` first and the same block is then imported so we should already have the right data, but it makes sense to output the hint prior to starting the block processing.

op-program also has a little trick up its sleeve where, unlike normal geth, when it returns a payload via `engine_getPayload` is also saves that block and its world state to the database. So when the created block is passed back to `engine_newPayload` it already exists in the database and the engine api doesn't bother reprocessing those transactions. We should make sure the hint is only updated after that check for the existing block in `newPayload` so we don't output it if we skip processing the block anyway.

Reply:
I updated the code to send on NewPayload as well. I think using the block number from the payload seems correct to you. The payload hasn't been imported yet at that point, right? So we're executing the block at `payload.BlockNumber`?

Also, I think we can send the hint regardless, even if we already have the block, because it's only fetched if a missing state/code key is requested, so extra hints shouldn't affect execution. (I think this highlights the distinction between hints and commands). It will also gracefully handle a reorg I believe because a reorg would mean the state nodes for the new block are missing causing the execution witness to be refetched.

Let me know if there's another reason to implement the logic to skip sending hint if a block already exists.

Reply:
Yes the block number for the payload is correct.  Sadly using block number will still break with potential reorgs. The game being played may be from _prior_ to the reorg so when op-node processes the re-org the block number will always refer to the wrong block. Admittedly the game being played will also have wound up being re-orged out and no longer exist so the consequences aren't significant but if cannon/op-program were already executing it would cause surprising results.  I can't imagine that ever happening in production, but test runs could be affected as they often work with the safe head and run outside of the context of a game on L1 so aren't necessarily affected by such a reorg.

I guess it doesn't really matter if we send a hint and then never actually request anything from it, just feels a bit weird.  It makes the code simpler though so I'm fine with it.